initial sht31 driver commit

jerryneedell opened the pull request:
This is the initial version of a driver for the sht31-d temperature/humidity sensor.

jerryneedell committed:
initial sht31 driver coomit

jerryneedell committed:
Update README.rst

jerryneedell committed:
Update README.rst

jerryneedell committed:
a few updates

jerryneedell committed:
remove README.md

jerryneedell committed:
fix up some syntax issues for lint

ladyada commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,153 @@
+    @property
+    def temperature(self):
+        """The measured relative humidity in percent."""
+        raw_temperature, raw_humidity = self._data()
```
i think for this, you should not replicate the math from temperature_and_relative_humidity but just call temperature_and_relative_humidity and return the first element

ladyada commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,153 @@
+    def relative_humidity(self):
+        """The measured relative humidity in percent."""
+        raw_temperature, raw_humidity = self._data()
+        return (100*(raw_humidity/65523))
```
ditto

ladyada commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,153 @@
+        else:
+            return False
+
+    @property
```
i think for _heater_ you can avoid having 'heater_status' and just have 'heater' be both a setter and getter property

jerryneedell committed:
remove redundant code for retreiving temp/humidity,  clean up heater control using setter

jerryneedell commented:
@ladyada Thanks for the review and comments. I think I have addressed them - let me know if I missed something or if you have other concerns.

ladyada commented:
heater looks excellent!

i had a change of heart after looking at other drivers - please remove temperature_and_relative_humidity(self) and just keep the temperature and humidity getter's - that way its cross-compatible with all our other T+H sensors. sorry & thank you :)

i'll also leave a note on the example py's inline

ladyada commented on examples/sht31_test.py:
```diff
@@ -0,0 +1,29 @@
+
+while True:
+    print(sensor.temperature_and_relative_humidity)
+    print(sensor.temperature)
```
For the example code, please mimic
https://github.com/adafruit/Adafruit_CircuitPython_SI7021/blob/master/examples/main.py
where we print text with each reading, just temp and humidity.

if you want to, maybe every 10 readings you can 'turn on the heater' for 1 second?

jerryneedell replied:
I removed sht31_test and did this in sht31_simpletest.py.

ladyada commented on examples/sht31_test.py:
```diff
@@ -0,0 +1,29 @@
+    sensor.heater=True
+    print(sensor.heater)
+    print(sensor.status)
+    sensor.reset
```
Hmm `sensor.reset` seems weird like this, make it a function if you don't actually assign anything?

jerryneedell replied:
changed to a function   - I did not include this in sht31_simpletest. Not sure if anyone will actually use it.

ladyada reviewed (changes requested):
example code review!

jerryneedell committed:
fix up example/sht31_simpletest.py - make sht31.reset a function

jerryneedell commented:
@ladyada No problem revising the getters - happy to make it work however you want it.

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+SHT31_HEATEREN = const(0x306D)
+SHT31_HEATERDIS = const(0x3066)
+
+def _crc(data):
```
Should be two blank lines before a `def`

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+                crc <<= 1
+    return crc
+
+class SHT31:
```
Same - need two blank lines before the class.

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+
+    @heater.setter
+    def heater(self, value=False):
+        if(value):
```
parens not needed `if value:`

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+
+    @property
+    def heater(self):
+        if(self.status & 0x2000):
```
if you decide to keep `if/else`, parens not needed: `if self.status & 0x2000:`

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+        time.sleep(.010)
+
+    @property
+    def heater(self):
```
This whole thing can read 
```
    @property
    def heater(self):
        return bool(self.status & 0x2000)
```
The if/else isn't needed.

kattni commented on adafruit_sht31.py:
```diff
@@ -108,36 +108,34 @@ def temperature_and_relative_humidity(self):
         """The measured relative humidity in percent."""
-        raw_temperature, raw_humidity = self._data()
-        return (-45 + (175 * (raw_temperature/65535)))
+        temperature, relative_humidity = self.temperature_and_relative_humidity
```
this can read 
```
    @property
    def temperature(self):
        """The measured relative humidity in percent."""
        return self.temperature_and_relative_humidity[0]
```

kattni commented on adafruit_sht31.py:
```diff
@@ -108,36 +108,34 @@ def temperature_and_relative_humidity(self):
         """The measured relative humidity in percent."""
-        raw_temperature, raw_humidity = self._data()
-        return (100*(raw_humidity/65523))
+        temperature, relative_humidity = self.temperature_and_relative_humidity
```
ditto except `return self.temperature_and_relative_humidity[1]`

kattni reviewed (changes requested):
Thanks so much for doing this!

I am going to try to dig up the sensor and test it, I have not tested it yet.

I'm having trouble getting inline reviews for the current commit, so there's a few more comments here. 

**You'll need to click "show outdated" in the list below**, as it thinks you've completed my suggestions because I applied them to the current commit.

You'll need to do a pylint disable invalid-name for this function, or rename it to something shorter - both are valid options, just be aware of the need. I've pasted it below with suggested changes: remove outside parens on the `return` and include whitespace around operators.
```
    @property
    def temperature_and_relative_humidity(self):
        """The measured temperature and relative humidity in percent."""
        raw_temperature, raw_humidity = self._data()
return -45 + (175 * (raw_temperature/65535)), 100 * (raw_humidity / 65523)
```

`status`, `heater`, and `reset` are all missing docstrings.

You may want to consider preparing this for ReadTheDocs. If you decide you want to, that's something I can help with. It involves extending the docstrings to include an image (the product image works), and a usage example where applicable. I believe it's something that will have to be done eventually, so if you're up for it, there's no better time than now.

jerryneedell commented:
@kattni Thanks for the review and comments. I work in the changes. FYI - the long named function temperature_and_relative_humidity has been removed in the current version.

kattni commented on examples/sht31_simpletest.py:
```diff
@@ -0,0 +1,25 @@
+while True:
+    print("\nTemperature: %0.1f C" % sensor.temperature)
+    print("Humidity: %0.1f %%" % sensor.relative_humidity)
+    loopcount+=1
```
need whitespace around the operator `loopcount += 1`

kattni commented on examples/sht31_simpletest.py:
```diff
@@ -0,0 +1,25 @@
+    loopcount+=1
+    time.sleep(2)
+    # every 10 passes turn on the heater for 1 second
+    if(loopcount==10):
```
no need for parens, need whitespace around operator `if loopcount == 10:`

kattni commented on examples/sht31_simpletest.py:
```diff
@@ -0,0 +1,25 @@
+    time.sleep(2)
+    # every 10 passes turn on the heater for 1 second
+    if(loopcount==10):
+        loopcount=0
```
whitespace around operator here and for both `sensor.heater` statements.

kattni reviewed (changes requested):
Tested successfully with the changes I suggested earlier. I'll want to test it with your final version.

I included some style updates to the example.

jerryneedell committed:
cleanup some lint issues

jerryneedell commented:
@kattni I think (hope) I addressed your comments. Let me know if I missed anything or you see something else.  I'll be happy to add the "readthedocs" stuff if you want.  Let me know what needs to be done. I also need to add some usage information to README.rst. Do you want that done as part of this PR or as a new one after this one is accepted. Either way is fine with me.

jerryneedell committed:
add missing doc stings

kattni commented on examples/sht31_simpletest.py:
```diff
@@ -0,0 +1,25 @@
+    # every 10 passes turn on the heater for 1 second
+    if loopcount == 10:
+        loopcount = 0
+        sensor.heater=True
```
whitespace around the operator `sensor.heater = True`

kattni commented on examples/sht31_simpletest.py:
```diff
@@ -0,0 +1,25 @@
+    if loopcount == 10:
+        loopcount = 0
+        sensor.heater=True
+        print("Sensor Heater status = ",sensor.heater)
```
remove space after `=`, add one space after comma: `print("Sensor Heater status =", sensor.heater)`

kattni commented on examples/sht31_simpletest.py:
```diff
@@ -0,0 +1,25 @@
+        sensor.heater=True
+        print("Sensor Heater status = ",sensor.heater)
+        time.sleep(1)
+        sensor.heater=False
```
whitespace around operator as above

kattni commented on examples/sht31_simpletest.py:
```diff
@@ -0,0 +1,25 @@
+        print("Sensor Heater status = ",sensor.heater)
+        time.sleep(1)
+        sensor.heater=False
+        print("Sensor Heater status = ",sensor.heater)
```
remove space after `=`, add space after comma as above

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,153 @@
+    """
+
+
+    def __init__(self, i2c_bus, address=SHT31_DEFAULT_ADDR):
```
You only need two spaces between a `def` and a different type of code. above this one, you don't need any empty lines and above each of the `def`s below, you only need one empty line.

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,153 @@
+        time.sleep(.010)
+
+
+    def _command(self, command):
```
only need one empty line above this, explained above. same for the rest of the `def`s in this code. (I won't add a comment to all of them)

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,153 @@
+
+
+    @property
+    def temperature(self):
```
replace `raw_humidity` in this function with `_`. it is an unused variable in this function, and is therefore not needed. `_` will serve to preserve the tuple, but make pylint happy about the variable issue. it should read as follows: `raw_temperature, _ = self._data()`

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,153 @@
+
+
+    @property
+    def relative_humidity(self):
```
same as above, but replace `raw_temperature` with `_`. Should read: `_, raw_humidity = self._data()`

kattni reviewed (changes requested):
This is looking great!

Looks like some of the changes I requested didn't show up for you (I think they showed up as completed when they weren't). I'm adding them again, if you're still not seeing them, let me know and I'll go over it with you.

There should be 4 in the library file, and 4 in the example file.

It's up to you when we do the RTD stuff. I think we'll want to discuss it on Discord because, while it's not really complicated, it can be a lot of info. Let me know how you'd like to do it - here or a new PR - and we'll get it worked out!

jerryneedell committed:
more cleanup for lint

jerryneedell commented:
@kattni OK - I think I got them this time, but please check me.   As for the documentation,  I would be fine with having this accepted then add the documentation updates as  a new PR but I'll do whatever you are comfortable with.

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+    @property
+    def temperature(self):
+        """The measured relative humidity in percent."""
+        raw_temperature,  _ = self._data()
```
You managed to get 2 spaces after the comma, needs to be 1.

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+    def temperature(self):
+        """The measured relative humidity in percent."""
+        raw_temperature,  _ = self._data()
+        return -45 + (175 * (raw_temperature/65535))
```
As long as you have one more thing to update, consider putting one space on either side of the `/` - this isn't making pylint unhappy, it's simply a readability thing.

tannewt replied:
I think it will make pylint unhappy. The repo wasn't enabled on Travis so it wasn't checking. I've turned it on now so it should catch new commits.

kattni commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+    def relative_humidity(self):
+        """The measured relative humidity in percent."""
+        _, raw_humidity = self._data()
+        return 100*(raw_humidity/65523)
```
Consider putting spaces around `*` and `/` - again not pylint, simply readability.

kattni reviewed (changes requested):
Super close! One more thing needed, and 2 suggested.

I'll go ahead and approve it once that's fixed up, and we can do RTD additions in a later PR. Thanks!

tannewt commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+
+    @property
+    def status(self):
+        """Read the sensor status."""
```
No need for "Read" it is the sensor status.

tannewt commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+    @property
+    def heater(self):
+        """Control the sensor internal heater."""
+        return bool(self.status & 0x2000)
```
Instead of casting using bool can you use equality instead? `return (self.status & 0x2000) != 0` I can never remember what numbers mean true and vice-versa. (Blame bash I think)

jerryneedell replied:
I like this!

tannewt commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+                i2c.readinto(data)
+        except OSError as error:
+            if error.args[0] not in ('I2C bus error', 19):  # errno 19 ENODEV
+                raise
```
Why hide the error that no device is found?

jerryneedell replied:
This was a hold over from the si7021 - It should not be here - I'll take it out.

tannewt replied:
I don't think you need the try except at all now. You are just re-raising it.

tannewt commented on adafruit_sht31.py:
```diff
@@ -0,0 +1,144 @@
+        if tcheck != _crc(data[:2]):
+            raise ValueError("temperature CRC mismatch")
+        if hcheck != _crc(data[3:5]):
+            raise ValueError("humidity CRC mismatch")
```
Please use RuntimeError instead. ValueError is usually raised when arguments passed in are not supported. Explanations of errors are here: https://docs.python.org/3/library/exceptions.html

tannewt reviewed (changes requested):
Great work on this Jerry! Just a style pass and expect pylint to nitpick too once you push another commit.

jerryneedell committed:
even more cleanup for lint - plus corrections to execption handling

jerryneedell committed:
missed one lint correction

jerryneedell committed:
add missing .files

kattni reviewed (approved):
Tested successfully with sensor. Looks great! Thanks so much for doing this! Great job!

jerryneedell committed:
remove trailing lines from sht31_simpletest.py fix whitepace in adafruit_sht31.py

tannewt commented on adafruit_sht31.py:
```diff
@@ -131,7 +130,7 @@ def heater(self, value=False):
     @property
     def status(self):
-        """Read the sensor status."""
+        """Return the Sensor status."""
```
Just `The sensor status`. From the outside status looks like a variable so it doesn't make sense to return anything.

jerryneedell committed:
revise exception handling - again - remove extra lines

jerryneedell committed:
fix sesnor.status doc string

kattni commented:
Looks great!
